// @maaz-jutt-utils/my-utils/src/multerCloudinary.js
import multer from "multer";
import { CloudinaryStorage } from "multer-storage-cloudinary";
import cloudinary from "cloudinary";

/**
 * Create a Multer upload instance with Cloudinary storage
 * @param {object} config - Cloudinary config
 * @param {string} config.cloudName - Cloudinary cloud name
 * @param {string} config.apiKey - Cloudinary API key
 * @param {string} config.apiSecret - Cloudinary API secret
 * @param {string} folder - Cloudinary folder name
 * @param {string} resourceType - "image" or "video"
 * @returns {object} multer upload instance
 */
export const createCloudinaryUpload = (
  { cloudName, apiKey, apiSecret },
  folder = "E-commerce/images",
  resourceType = "image"
) => {
  // Configure Cloudinary
  cloudinary.v2.config({
    cloud_name: cloudName,
    api_key: apiKey,
    api_secret: apiSecret,
  });

  // Create Cloudinary storage
  const storage = new CloudinaryStorage({
    cloudinary: cloudinary.v2,
    params: async (req, file) => ({
      folder,
      resource_type: resourceType,
      allowed_formats: ["png", "jpg", "jpeg", "mp4", "mp3"],
    }),
  });

  return multer({ storage });
};
